contracts/TradeTypesLib2.sol

```diff
@@ -0,0 +1,134 @@
+/*
+ * Hubii Nahmii
+ *
+ * Compliant with the Hubii Nahmii specification v0.12.
+ *
+ * Copyright (C) 2017-2018 Hubii AS
+ */
+
+pragma solidity >=0.4.25 <0.6.0;
+
+import {MonetaryTypesLib} from "./MonetaryTypesLib.sol";
+import {NahmiiTypesLib} from "./NahmiiTypesLib.sol";
+
+/**
+ * @title     TradeTypesLib
+ * @dev       Data types centered around trade
+ */
+library TradeTypesLib2 {
+    //
+    // Enums
+    // -----------------------------------------------------------------------------------------------------------------
+    enum CurrencyRole {Intended, Conjugate}
+    enum LiquidityRole {Maker, Taker}
+    enum Intention {Buy, Sell}
+    enum TradePartyRole {Buyer, Seller}
+
+    //
+    // Structures
+    // -----------------------------------------------------------------------------------------------------------------
+    struct Operator {
+        uint256 id;
+        string data;
+    }
+
+    struct OrderPlacement {
+        Intention intention;
+
+        int256 amount;
+        NahmiiTypesLib.IntendedConjugateCurrency currencies;
+        int256 rate;
+
+        NahmiiTypesLib.CurrentPreviousInt256 residuals;
+    }
+
+    struct OrderParty {
+        uint256 nonce;
+        address wallet;
+
+        NahmiiTypesLib.IntendedConjugateCurrency balances;
+
+        string data;
+    }
+
+    struct Order {
+        OrderParty party;
+
+        OrderPlacement placement;
+
+        NahmiiTypesLib.WalletOperatorSeal seals;
+        uint256 blockNumber;
+
+        Operator operator;
+    }
+
+    struct OrderCancellation {
```

Comment:
How do we associate an OrderCancellation with the order to cancel? 

For instance, what happens if A places two identical buy orders, and wants to cancel one of them?

Reply:
Would it make sense to use the wallet's hash of the order to specify which order to try to delete? This hash will be unique, even for identical orders, due to the "sender data" property we use against reply attacks.

Reply:
I reckon we have (at least) two alternatives. The wallet hash is one. An alternative one would be the addition of wallet order nonce, i.e. a different series than wallet driip nonce.

I suggest this be considered more in depth in the context of #419.

Reply:
Adding more nonces seems like a counterproductive way to identify orders. Accepted orders (signed by the operator) will have a unique operator hash due to the unique constraint on the wallet "data" property (which is included in the wallet hash). Thus I would claim that an operator hash is for all intents and purposes "the id" of an order. 
Using the wallet hash doesnt make sense since that would only identify the order _request_ to the system and not an actual accepted order - and only accepted orders can be cancelled after all :)